operator/pkg/overlays/postrenderer.go

```diff
@@ -0,0 +1,85 @@
+package overlays
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+
+	"github.com/stackrox/k8s-overlay-patch/pkg/patch"
+	"github.com/stackrox/k8s-overlay-patch/pkg/types"
+	"github.com/stackrox/rox/operator/apis/platform/v1alpha1"
+	"helm.sh/helm/v3/pkg/postrender"
+	ctrlClient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type OverlayPostRenderer struct {
+	obj              ctrlClient.Object
+	defaultNamespace string
+}
+
+func NewOverlayPostRenderer(obj ctrlClient.Object, defaultNamespace string) *OverlayPostRenderer {
+	return &OverlayPostRenderer{
+		obj:              obj,
+		defaultNamespace: defaultNamespace,
+	}
+}
+
+var _ postrender.PostRenderer = &OverlayPostRenderer{}
+
+type crdWithOverlays struct {
+	Spec overlaySpec `json:"spec"`
+}
+
+type overlaySpec struct {
+	Overlays []*v1alpha1.K8sObjectOverlay `json:"overlays"`
+}
+
+func (o OverlayPostRenderer) Run(renderedManifests *bytes.Buffer) (modifiedManifests *bytes.Buffer, err error) {
+	objBytes, err := json.Marshal(o.obj)
+	if err != nil {
+		fmt.Println("error marshalling object", err)
+		return nil, err
+	}
+
+	var wo crdWithOverlays
+	err = json.Unmarshal(objBytes, &wo)
+	if err != nil {
+		fmt.Println("error unmarshalling rendered manifests", err)
+		return nil, err
+	}
+
+	if len(wo.Spec.Overlays) == 0 {
+		return renderedManifests, nil
+	}
+
+	patched, err := patch.YAMLManifestPatch(renderedManifests.String(), o.defaultNamespace, mapOverlays(wo.Spec.Overlays))
```

Comment:
In an offline chat @ludydoo:
- clarified that this does a deserialize/serialize round-trip internally for all the manifests to apply the overlay.
- reminded me that helm-operator library [already does](https://github.com/stackrox/helm-operator/blob/08cc94506b32b820d75179190dd1f3fc1c287bd1/pkg/client/postrenderer.go#L115) a deserialize-serialize round-trip for all manifests in order to set the owner references/annotations

Combined with a third final deserialize/serialize round trip to push out the objects through the kube client, I'm concerned that this might be the straw that broke the camel's back. Probably not for a typical single-CR-per-cluster user, but certainly a concern for a bulk use case such as the ACS cloud service.

We should definitely do some benchmarks to make sure this is actually usable with many large CRs.

When thinking about how this could be optimized, I took a look at the related [helm code](https://github.com/helm/helm/blob/14d363669f8cc257f0413995d42237e673422d81/pkg/action/install.go#L289-L311):
- line 289 performs the render and runs the post-renderers
- line 292 the output is saved in the release object
- line 305 deserializes the manifests into objects
- line 311 walks the objects to adjust their annotations and labels (the changes are **not** reflected in the manifests saved in the release, not sure if this is intentional)
- further below, the objects are sent to kube API

The upgrade code works similarly.

I wonder whether it would be feasible to adjust the helm API to (in addition to accepting a custom post-renderer) also accept a custom visitor that could be used to modify already deserialized objects.

If not, then perhaps we could make changes such that a single additional parsing/serialization round  is performed (rather then two of them).

Reply:
I performed a benchmark test, here are the test info & the result

Manifests:
- 100 deployments, 3 pods each
- Each pod has...
  - 20 volume mounts
  - 20 env vars
  - 20 commands
  - 3 ports
- Total number of lines: 42 000

For comparison, the `Central` manifests currently have 1634 lines of yaml. So this manifest is ~25x bigger. 

### Result for 100 patches in the `Central`
```
BenchmarkOverlayPostRenderer_Run-10    	       6	 178899806 ns/op (~0.17s)
```

### Result for 1 patch in `Central`
```
BenchmarkOverlayPostRenderer_Run-10    	      14	  82796295 ns/op (~0.08s)
```

### Running for 3 patches in a smaller manifest with ~1600 lines of yaml
```
BenchmarkOverlayPostRenderer_Run-10    	     171	   6971124 ns/op (~0.006s)
```

This seems like it will have a minor impact on performance